### PR TITLE
Re-center the map when zooming with active scale lock

### PR DESCRIPTION
**Pull request by jdugge**
## Description

When zooming with the mouse wheel or the Zoom map tool while the scale is locked the zoom is not relative to the mouse / zoom box position, but relative to the center of the screen (see #30023).

This PR fixes this behavior by adding a new optional `center` argument to the 
`QgsMapSettings::setMagnificationFactor` and `QgsMapCanvas::setMagnificationFactor` functions, which is then used to scale the extent rectangle.

### CODE REVIEW

File: src/app/qgisapp.cpp [hunk 1]
```diff
@@ -3683,7 +3683,7 @@ void QgisApp::createStatusBar()
   mMagnifierWidget->setObjectName( QStringLiteral( "mMagnifierWidget" ) );
   mMagnifierWidget->setFont( statusBarFont );
   connect( mMapCanvas, &QgsMapCanvas::magnificationChanged, mMagnifierWidget, &QgsStatusBarMagnifierWidget::updateMagnification );
-  connect( mMagnifierWidget, &QgsStatusBarMagnifierWidget::magnificationChanged, mMapCanvas, &QgsMapCanvas::setMagnificationFactor );
+  connect( mMagnifierWidget, &QgsStatusBarMagnifierWidget::magnificationChanged, mMapCanvas, [ = ]( double factor ) { mMapCanvas->setMagnificationFactor( factor ); } );
```
Comment (3nids) [inline src/app/qgisapp.cpp:3686]: why is this change required since there is a default value in the slot?

Comment (jdugge) [reply src/app/qgisapp.cpp:3686]: The Qt5 `connect` syntax [does not support default arguments in slots](https://wiki.qt.io/New_Signal_Slot_Syntax#Default_arguments_in_slot), unfortunately. I used a lambda function to work around this, which seems to be the standard way to handle this ([example 1](https://artandlogic.com/2013/09/qt-5-and-c11-lambdas-are-your-friend/), [example 2](https://stackoverflow.com/a/40598273/1456857)).

**Comment by nyalldawson** [comment]
@3nids what's your thoughts? Can we merge?

**Comment by jdugge** [comment]
@3nids: Is there anything I should change so this PR can be merged?

**Review by nyalldawson — APPROVED** [review]
